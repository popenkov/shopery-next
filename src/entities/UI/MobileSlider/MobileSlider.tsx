'use client';

import { DetailedHTMLProps, HTMLAttributes, PropsWithChildren } from 'react';
import { Swiper, SwiperSlide } from 'swiper/react';
import 'swiper/css';

import cls from './MobileSlider.module.scss';

interface MobileSliderProps<T>
  extends DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>,
    PropsWithChildren<{
      data?: T[];
      render: (child: T) => JSX.Element;
    }> {}

export const MobileSlider = <T,>({
  data,
  render,
  ...restProps
}: MobileSliderProps<T>) => {
  if (!data?.length) {
    return null;
  }

  return (
    <div className={cls.slider} {...restProps}>
      <Swiper
        spaceBetween={20}
        slidesPerView={'auto'}
        loop
        className={cls.swiper}
      >
        {data.map((item, index) => {
          return <SwiperSlide key={index}>{render(item)}</SwiperSlide>;
        })}
      </Swiper>
    </div>
  );
};
