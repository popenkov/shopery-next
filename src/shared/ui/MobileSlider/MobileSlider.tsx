import React from 'react';

export const MobileSlider = () => {
  return <div>MobileSlider</div>;
};
