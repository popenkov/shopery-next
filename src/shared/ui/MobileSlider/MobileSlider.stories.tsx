import type { Meta, StoryObj } from '@storybook/react';

// eslint-disable-next-line
import { Banner, TBanner } from '@/entities/Banner';
// eslint-disable-next-line
import { getBannerData } from '@/widgets/Banners';

import { MobileSlider } from './MobileSlider';

const meta = {
  title: 'shared/MobileSlider',
  component: MobileSlider,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  args: {},
  argTypes: {},
} satisfies Meta<typeof MobileSlider>;

export default meta;
type TStory = StoryObj<typeof MobileSlider<TBanner>>;

const { banners } = getBannerData();

export const MobileSliderComponent: TStory = {
  decorators: [
    (Story) => (
      <div>
        <Story />
      </div>
    ),
  ],
  args: {
    data: banners,
    render: (child: TBanner) => <Banner {...child} />,
  },
};
