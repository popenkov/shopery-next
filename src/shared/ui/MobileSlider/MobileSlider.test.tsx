import React, { ReactNode } from 'react';
import { render } from '@testing-library/react';

import { MobileSlider } from './MobileSlider';
jest.mock('swiper/react', () => ({
  Swiper: ({ children }: { children: ReactNode }) => (
    <div data-testid="swiper-testid">{children}</div>
  ),
  SwiperSlide: ({ children }: { children: ReactNode }) => (
    <div data-testid="swiper-slide-testid">{children}</div>
  ),
}));
const data = [{ id: 1 }, { id: 2 }, { id: 3 }];
describe('MobileSlider component', () => {
  it('renders correctly', () => {
    const renderSlides = (item: { id: number }) => <div>{item.id}</div>;
    const { container } = render(<MobileSlider data={data} render={renderSlides} />);
    expect(container).toMatchSnapshot();
  });

  it('renders swiper correctly', () => {
    const renderSlides = (item: { id: number }) => <div>{item.id}</div>;
    const { container } = render(<MobileSlider data={data} render={renderSlides} />);
    const swiper = container.querySelector('.swiper');
    expect(swiper).toBeInTheDocument();
  });

  it('renders swiper slides correctly', () => {
    const renderSlides = (item: { id: number }) => <div>{item.id}</div>;
    const { container } = render(<MobileSlider data={data} render={renderSlides} />);
    const swiperSlides = container.querySelectorAll('.swiper-slide');
    expect(swiperSlides).toHaveLength(3);
  });

  it('renders children correctly', () => {
    const renderSlides = (item: { id: number }) => <div>{item.id}</div>;
    const { getAllByText } = render(<MobileSlider data={data} render={renderSlides} />);
    const children = getAllByText(/1|2|3/);
    expect(children).toHaveLength(3);
  });

  it('does not render when data is empty', () => {
    const data: [] = [];
    const renderSlides = (item: { id: number }) => <div>{item.id}</div>;
    const { queryByText } = render(<MobileSlider data={data} render={renderSlides} />);
    expect(queryByText(/1|2|3/)).not.toBeInTheDocument();
  });
});
