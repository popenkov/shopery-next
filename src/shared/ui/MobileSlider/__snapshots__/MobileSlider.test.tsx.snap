// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobileSlider component renders correctly 1`] = `
<div>
  <div
    class="slider"
  >
    <div
      data-testid="swiper-testid"
    >
      <div
        data-testid="swiper-slide-testid"
      >
        <div>
          1
        </div>
      </div>
      <div
        data-testid="swiper-slide-testid"
      >
        <div>
          2
        </div>
      </div>
      <div
        data-testid="swiper-slide-testid"
      >
        <div>
          3
        </div>
      </div>
    </div>
  </div>
</div>
`;
